### PR TITLE
[CELEBORN-487][PERF] ShuffleClientSide support blacklist to avoid client side timeout in same worker multiple times

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
During test if three worker one worker was stuck, 

Before improve 
image

It was caused by for one shuffle, there may have not only one partition on the stuck worker, then it will cause the shuffle client delay several timeout times during push data.
Also if for merged data, it may also push to the stuck node even after we have revive the partition location.
 

After this pr 
截屏2023-03-30 下午3 04 31

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture @RexXiong @FMX @pan3793

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 1]
```diff
@@ -1196,6 +1242,12 @@ public void onFailure(Throwable e) {
                 remainReviveTimes,
                 e);
             if (!mapperEnded(shuffleId, mapId, attemptId)) {
+              int tmpRemainReviveTimes = remainReviveTimes;
+              // For blacklisted partition location, Celeborn should not use retry quota.
+              if (!isPushBlacklisted(cause)) {
+                tmpRemainReviveTimes = tmpRemainReviveTimes - 1;
+              }
+              int finalRemainReviveTimes = tmpRemainReviveTimes;
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1250]: finalRemainReviveTimes is unnecessary, just use tmpRemainReviveTimes

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -1504,6 +1505,14 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(false)
 
+  val SHUFFLE_CLIENT_BLACKLIST_ENABLED: ConfigEntry[Boolean] =
+    buildConf("celeborn.shuffle.client.blacklist.enable")
```
Comment (waitinfuture) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1509]: celeborn.shuffle.client.blacklist.enable => celeborn.client.blacklist.enabled

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -1504,6 +1505,14 @@ object CelebornConf extends Logging {
...
+      .categories("client")
+      .doc("Whether to enable shuffle client blacklist.")
```
Comment (waitinfuture) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1511]: Whether to enable shuffle client-side blacklist of workers

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 4]
```diff
@@ -1460,6 +1519,11 @@ protected boolean mapperEnded(int shuffleId, int mapId, int attemptId) {
         && mapperEndMap.get(shuffleId).contains(Utils.makeMapKey(shuffleId, mapId, attemptId));
   }
 
+  private boolean isPushBlacklisted(StatusCode cause) {
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1522]: pushBlacklisted

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java [hunk 5]
```diff
@@ -213,18 +217,30 @@ private void submitRetryPushData(
...
-          TransportClient client =
-              dataClientFactory.createClient(newLoc.getHost(), newLoc.getPushPort(), partitionId);
-          NettyManagedBuffer newBuffer = new NettyManagedBuffer(Unpooled.wrappedBuffer(body));
-          String shuffleKey = Utils.makeShuffleKey(applicationId, shuffleId);
-
-          PushData newPushData =
-              new PushData(MASTER_MODE, shuffleKey, newLoc.getUniqueId(), newBuffer);
-          client.pushData(newPushData, pushDataTimeout, wrappedCallback);
+        // If shuffleClientBlacklistEnabled = false, blacklist should be empty.
+        if (blacklist.contains(newLoc.hostAndPushPort())) {
```
Comment (waitinfuture) [inline client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:221]: Better to extract L221-L227 to a common method like checkBlacklist()

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java (hunk 1 above)
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1250]: 截屏2023-04-02 上午9 36 49

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:1509]: > celeborn.client.blacklist.enabled

How about  celeborn.client.push.blacklist.enabled? later will also do similar logical in read

Comment (waitinfuture) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala]: LGTM

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java (hunk 1 above)
Comment (waitinfuture) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java:1252]: Got it, thanks

File: client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java (hunk 5 above)
Comment (AngersZhuuuu) [reply client/src/main/java/org/apache/celeborn/client/ShuffleClientImpl.java]: How about current?

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture any more suggestion

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks!

**Comment by cfmcgrady** [comment]
Given that we already possess a `WorkerStatusTracker` to monitor the status of workers, we can leverage the existing functionality of the `WorkerStatusTracker` to streamline the implementation of the blacklist logic? @AngersZhuuuu @waitinfuture

**Comment by AngersZhuuuu** [comment]
Yea, we have thought about such things but plat to do this later. If you want to do it, just do it.

**Comment by waitinfuture** [comment]
Hi @cfmcgrady , thanks for your interest on the blacklist functionality! Currently the design is quite complicated, I try to explain.

- There are three roles who have blacklist: Master, LifecycleManager(in Driver of Spark, and JobMaster of Flink), and ShuffleClientImpl(in Executor of Spark, and TaskManager of Flink). The three roles have different ways to decide which worker is blacklisted.
- For Master, the only way to put a worker into blacklist is when a worker tells that it has no available disks. See ```AbstractMetaManager.updateWorkerHeartbeatMeta```. Notice that if a worker has heartbeat timeout or is stopped without graceful shutdown, the Master simply removes the worker from the blacklist, see ```AbstractMetaManager.updateWorkerLostMeta```. If a worker is graceful shutdown, it is put in the ```shutdownWorkers```
- For LifecycleManager, there are three ways to put a worker into local blacklist.
  1. Reported by worker through ```Revive``` and ```PartitionSplit```, and only certain causes will consider as should be blacklisted, for example ```PUSH_DATA_WRITE_FAIL_MASTER/PUSH_DATA_CREATE_CONNECTION_FAIL_MASTER```, see ```ChangePartitionManager.handleRequestPartitions```
  2. When LifecycleManager encounters RPC failure to worker, for example commit files failure and reserve slots failure, see ```WorkerStatusTracker.recordWorkerFailure```
  3. Refreshed by the Master through heartbeat response. ```HeartbeatFromApplicationResponse``` will return blacklist/shutdownWorkers/unkownworkers from Master. The fresh logic is in ```WorkerStatusTracker.handleHeartbeatResponse```.
- For ShuffleClientImpl, the reason to put a worker into local blacklist is when push data fails, and only certain causes will be put into blacklist, see ```ShuffleClientImpl.revive```. The cause will be passed to LifecycleManager through Revive. The local blacklist is refreshed according to ```ChangeLocationResponse``` and ```RegisterShuffleResponse```.

So, Master's blacklist is from worker status, ShuffleClientImpl's blacklist is from itself, LifecycleManager's blacklist is from Master/ShuffleClientImpl/LifecycleManager. And the priority is Master > LifecycleManager > ShuffleClientImpl.

I'm not sure if all three roles' blacklist logic can be unified by ```WorkerStatusTracker```, although it'll be great if it can.

BTW, ShuffleClientImpl also has blacklist for shuffle read.

**Comment by pan3793** [comment]
It worth to write a blog to explain how the blacklist mechanism works :)

**Comment by waitinfuture** [comment]
> It worth to write a blog to explain how the blacklist mechanism works :)

Yeah, and lots of other modules : )

**Comment by cfmcgrady** [comment]
thanks for the detailed explanation.

**Pull request merged**